File: nodecore-grpc/src/main/java/nodecore/api/grpc/utilities/ChannelBuilder.java
// VeriBlock NodeCore
// Copyright 2017-2020 Xenios SEZC
// All rights reserved.
// https://www.veriblock.org
// Distributed under the MIT software license, see the accompanying
// file LICENSE or http://www.opensource.org/licenses/mit-license.php.

package nodecore.api.grpc.utilities;

import io.grpc.*;
import io.grpc.netty.shaded.io.grpc.netty.GrpcSslContexts;
import io.grpc.netty.shaded.io.grpc.netty.NegotiationType;
import io.grpc.netty.shaded.io.grpc.netty.NettyChannelBuilder;
import io.grpc.stub.MetadataUtils;
import nodecore.api.grpc.AdminRpcConfiguration;
import nodecore.api.grpc.Constants;

import javax.net.ssl.SSLException;
import java.io.File;

public class ChannelBuilder {
    private final AdminRpcConfiguration configuration;

    public ChannelBuilder(AdminRpcConfiguration configuration) {
        this.configuration = configuration;
    }

    public ManagedChannel buildManagedChannel() throws SSLException {
        if (configuration.isSsl()) {
            return buildTlsManagedChannel();
        }

        return buildPlainTextManagedChannel();
    }

    public Channel attachPasswordInterceptor(Channel inner) {
        String password = configuration.getNodeCorePassword();
        Metadata headers = new Metadata();
        if(password != null) {
            headers.put(Constants.RPC_PASSWORD_HEADER_NAME, password);
        }
        ClientInterceptor clientInterceptor = MetadataUtils.newAttachHeadersInterceptor(headers);
        return ClientInterceptors.intercept(inner, clientInterceptor);
    }

    private ManagedChannel buildTlsManagedChannel() throws SSLException {
        File certChainFile = new File(configuration.getCertificateChainPath());
        return NettyChannelBuilder.forAddress(configuration.getNodeCoreHost(), configuration.getNodeCorePort())
                .sslContext(GrpcSslContexts.forClient().trustManager(certChainFile).build())
                .negotiationType(NegotiationType.TLS)
                .build();
    }

    private ManagedChannel buildPlainTextManagedChannel() {
        return NettyChannelBuilder.forAddress(configuration.getNodeCoreHost(), configuration.getNodeCorePort())
                .usePlaintext()
                .build();
    }
}

File: pop-miners/veriblock-pop-miners-common/src/main/kotlin/org/veriblock/miners/pop/service/ChannelBuilder.kt
// VeriBlock PoP Miner
// Copyright 2017-2020 Xenios SEZC
// All rights reserved.
// https://www.veriblock.org
// Distributed under the MIT software license, see the accompanying
// file LICENSE or http://www.opensource.org/licenses/mit-license.php.
package org.veriblock.miners.pop.service

import io.grpc.Channel
import io.grpc.ClientInterceptors
import io.grpc.ManagedChannel
import io.grpc.Metadata
import io.grpc.netty.shaded.io.grpc.netty.GrpcSslContexts
import io.grpc.netty.shaded.io.grpc.netty.NegotiationType
import io.grpc.netty.shaded.io.grpc.netty.NettyChannelBuilder
import io.grpc.stub.MetadataUtils
import org.veriblock.core.utilities.createLogger
import org.veriblock.miners.pop.Constants
import org.veriblock.miners.pop.common.NodeCoreRpcConfig
import java.io.File

private val logger = createLogger {}

class ChannelBuilder(
    private val config: NodeCoreRpcConfig
) {
    fun buildManagedChannel(): ManagedChannel {
        logger.info(
            "Connecting to NodeCore at {}:{} {}", config.host, config.port,
            if (config.ssl) "over SSL" else ""
        )
        return if (config.ssl) {
            buildTlsManagedChannel()
        } else {
            buildPlainTextManagedChannel()
        }
    }

    fun attachPasswordInterceptor(inner: Channel?): Channel {
        val headers = Metadata()
        val password = config.password
        if (password != null) {
            headers.put(Constants.RPC_PASSWORD_HEADER_NAME, password)
        }
        val clientInterceptor = MetadataUtils.newAttachHeadersInterceptor(headers)
        return ClientInterceptors.intercept(inner, clientInterceptor)
    }

    private fun buildTlsManagedChannel(): ManagedChannel {
        if (config.certificateChainPath == null) {
            error("'nodecore.rpc.certificateChainPath' is not configured")
        }
        val certChainFile = File(config.certificateChainPath)
        return NettyChannelBuilder
            .forAddress(config.host, config.port)
            .sslContext(GrpcSslContexts.forClient().trustManager(certChainFile).build())
            .negotiationType(NegotiationType.TLS)
            .build()
    }

    private fun buildPlainTextManagedChannel(): ManagedChannel {
        return NettyChannelBuilder
            .forAddress(config.host, config.port)
            .usePlaintext()
            .build()
    }
}
